modules/smtp/modules.php

```diff
@@ -939,6 +946,132 @@ protected function output() {
+                            }
+                            if (new DateTime($msg_headers['X-Schedule']) <= new DateTime()) {
+                                $profile = Hm_Profiles::get($msg_headers['X-Profile-ID']);
+                                if (!$profile) {
+                                    $profiles = Hm_Profiles::search('server', $imap_details['server']);
+
+                                    if (!$profiles) {
+                                        Hm_Debug::add(sprintf('ERRCannot find profiles corresponding with IMAP server: %s', $imap_details['server']));
+                                        continue;
+                                    }
+                                    $profile = $profiles[0];
+                                }
+
+                                $smtp = Hm_SMTP_List::connect($profile['smtp_id'], false);
+
+                                if (smtp_authed($smtp)) {
+                                    if (isset($msg_headers['X-Delivery'])) {
+                                        $from_params = 'RET=HDRS';
+                                        $recipients_params = 'NOTIFY=SUCCESS,FAILURE';
+                                    } else {
+                                        $from_params = '';
+                                        $recipients_params = '';
+                                    }
+                                    $recipients = [];
+                                    foreach (['To', 'Cc', 'Bcc'] as $fld) {
+                                        if (array_key_exists($fld, $msg_headers)) {
+                                            $recipients = array_merge($recipients, Hm_MIME_Msg::find_addresses($msg_headers[$fld]));
+                                        }
+                                    }
+                                    $msg_content = $imap->get_message_content($msg['uid'], 0);
+                                    $from = process_address_fld($msg_headers['From']);
+
+                                    $err_msg = $smtp->send_message($from[0]['email'], $recipients, $msg_content, $from_params, $recipients_params);
+
+                                    if (!$err_msg) {
+                                        if ($imap->message_action('DELETE', [$msg['uid']])) {
+                                            $imap->message_action('EXPUNGE', [$msg['uid']]);
+                                        }
+                                        save_sent_msg($this, $server_id, $imap, $imap_details, $msg_content, $msg['uid'], false);
+                                        $scheduled_msg_count--;
+                                    }
+                                }
+                            }
+                        } catch (Exception $e) {
+                            Hm_Debug::add(sprintf('ERRCannot send message: %s', $msg_headers['subject']));
+                        if (send_scheduled_message($this, $imap, $msg, $server_id)) {
```

Comment:
If you retry sending in some other way, combine the above sending with the one in this method, so we don't repeat so much code.